discord/utils.py

```diff
@@ -573,64 +589,132 @@ def get(iterable: Iterable[T], **attrs: Any) -> T | None:
+        AppEmoji,
+        Client,
+        Guild,
+        Member,
+        User,
+        abc,
+        emoji,
+    )
 
-    Examples
-    --------
+    string_to_type = {
+        "channel": abc.GuildChannel,
+        "member": Member,
+        "user": User,
+        "guild": Guild,
+        "emoji": emoji._EmojiTag,
+        "appemoji": AppEmoji,
+    }
+
+    if attr is not MISSING or id is not MISSING or isinstance(object_type, str):
+        warn_deprecated(
+            name="get_or_fetch(obj, attr='type', id=...)",
+            instead="get_or_fetch(obj, object_type=Type, object_id=...)",
+            since="2.7",
+        )
 
-    Getting a guild from a guild ID: ::
+        deprecated_attr = attr if attr is not MISSING else object_type
+        deprecated_id = id if id is not MISSING else object_id
+
+        if isinstance(deprecated_attr, str):
+            mapped_type = string_to_type.get(deprecated_attr.lower())
+            if mapped_type is None:
+                raise InvalidArgument(
+                    f"Unknown type string '{deprecated_attr}' used. Please use a valid object class like `discord.Member` instead."
+                )
+            object_type = mapped_type
+        elif isinstance(deprecated_attr, type):
+            object_type = deprecated_attr
+        else:
+            raise TypeError(
+                f"Invalid `attr` or `object_type`: expected a string or class, got {type(deprecated_attr).__name__}."
+            )
 
-        guild = await utils.get_or_fetch(client, 'guild', guild_id)
+        object_id = deprecated_id
 
-    Getting a channel from the guild. If the channel is not found, return None: ::
+    if object_type is MISSING or object_id is MISSING:
+        raise TypeError("required parameters: `object_type` and `object_id`.")
 
-        channel = await utils.get_or_fetch(guild, 'channel', channel_id, default=None)
-    """
-    getter = getattr(obj, f"get_{attr}")(id)
-    if getter is None:
+    if issubclass(object_type, (Member, User, Guild)):
+        attr = object_type.__name__.lower()
+    elif issubclass(object_type, emoji._EmojiTag):
+        attr = "emoji"
+    elif issubclass(object_type, abc.GuildChannel):
+        attr = "channel"
+    else:
+        raise InvalidArgument(
+            f"Class {object_type.__name__} cannot be used with discord.{type(obj).__name__}.get_or_fetch()"
+        )
+    if isinstance(obj, Guild) and object_type is User:
+        raise InvalidArgument(
+            "Guild cannot get_or_fetch discord.User. Use Client instead."
+        )
+    elif isinstance(obj, Client) and object_type is Member:
+        raise InvalidArgument("Client cannot get_or_fetch Member. Use Guild instead.")
+
+    getter = getattr(obj, f"get_{attr}", None)
+    if getter:
+        result = getter(object_id)
+        if result is not None:
+            return result
+
+    fetcher = getattr(obj, f"fetch_{attr}", None) or getattr(
+        obj, f"_fetch_{attr}", None
+    )
```

Comment:
instead of all this we could store a mapping, like:
```py
mapping = {
    Member: (Guild.get_member, Guild.fetch_member),
    ...
}
getter, fetcher = mapping[object_type]
result = getter(obj, object_id)
if not result:
    # try/except obviously
    result = await fetcher(obj, object_id)
return result